### PR TITLE
Allow required arguments after optional arguments

### PR DESCRIPTION
**Pull request by willkroboth**
This PR is me implementing what I was trying to explain [here](https://discord.com/channels/745416925924032523/745419648970784821/1060603371159044157) in discord so I can explain it better.

Currently, if you put a required argument after an optional argument, you get an `OptionalArgumentException`. However, I think this example command makes a good case for changing that:
```java
new CommandAPICommand("rate")
	.withArguments(
		new StringArgument("topic").setOptional(true),
		new IntegerArgument("rating", 0, 10),
		new PlayerArgument("target").setOptional(true)
	)
	.executes(info -> {
		String topic = (String) info.args().get("topic");
		if(topic == null) {
			// Honestly not sure how to justify topic being optional here,
			// causing there to be a required argument after an optional one,
			// but I'm sure an example exists, right?
			info.sender().sendMessage("You didn't give a rating");
			return;
		}

		// We know this is not null because rating is required if topic is given
		int rating = (int) info.args().get("rating");

		// The target player is optional, so give it a default here
		CommandSender target = (CommandSender) info.args().getOrDefault("target", info.sender());

		target.sendMessage("Your " + topic + " was rated: " + rating + "/10");
	})
	.register();
```

This command can be executed in the following ways:
```
/rate
/rate <topic> <rating>
/rate <topic> <rating> <target>
```

Putting the required `<rating>` argument after the optional `<topic>` argument indicates that if a `topic` is given, a `rating` is required.

To make this work was actually very simple, just a small change to the old system. `AbstractCommandAPICommand#getArgumentsToRegister` now looks like this:
```java
private List<Argument[]> getArgumentsToRegister(Argument[] argumentsArray) {
	List<Argument[]> argumentsToRegister = new ArrayList<>();

	// Create a List of arrays that hold arguments to register split based on where optional arguments are
	List<Argument> currentCommand = new ArrayList<>(argumentsArray.length);
	for(Argument argument : argumentsArray) {
		if(argument.isOptional()) argumentsToRegister.add((Argument[]) currentCommand.toArray(new AbstractArgument[0]));
		currentCommand.add(argument);
	}
	// All arguments are also valid and not added by above loop
	argumentsToRegister.add(argumentsArray);
	return argumentsToRegister;
}
```

### TODO before merge:
- Get the idea approved
- Remove `OptionalArgumentException`? (I don't think it is used anymore)
- [ ] Update documentation
- [ ] Update `commandapi-bukkit-plugin-test`
- [ ] Remove the example command from CommandAPIMain

### CODE REVIEW

File: commandapi-platforms/commandapi-bukkit/commandapi-bukkit-plugin/src/main/java/dev/jorel/commandapi/CommandAPIMain.java [hunk 1]
```diff
@@ -125,5 +129,32 @@ public void onLoad() {
...
+				new IntegerArgument("rating", 0, 10),
+				new PlayerArgument("target").setOptional(true)
+			)
+			.executes(info -> {
+				String topic = (String) info.args().get("topic");
+				if(topic == null) {
+					// Honestly not sure how to justify topic being optional here,
+					// causing there to be a required argument after an optional one,
+					// but I'm sure an example exists, right?
+					info.sender().sendMessage("You didn't give a rating");
```
Comment (JorelAli) [inline commandapi-platforms/commandapi-bukkit/commandapi-bukkit-plugin/src/main/java/dev/jorel/commandapi/CommandAPIMain.java:146]: There is actually a use case for this! A help message stating the usage for this command. If a player types `/rate`, you could feasibly follow up with sending a message to the user stating how to use `/rate`

**Comment by JorelAli** [comment]
> ```java
> // Honestly not sure how to justify topic being optional here,
> // causing there to be a required argument after an optional one,
> // but I'm sure an example exists, right?
> ```

There is actually a use case for this! A help message stating the usage for this command. If a player types `/rate`, you could feasibly follow up with sending a message to the user stating how to use `/rate`

**Comment by DerEchtePilz** [comment]
Writing this here again (this comment comes from [this](https://discord.com/channels/745416925924032523/745419648970784821/1060624350232973443) comment on discord):
What if we make this possible by "linking" arguments together?
What I mean by this is that if we made that command /rate <topic> <rating> <player> with the syntax Will Kroboth suggested, that <topic> and <rating> are basically one argument when checking optional argument constraints but when we are actually building that argumentsToRegister list it would be possible to register them individually making that syntax possible.

That wouldn't invalidate the `OptionalArgumentException` and would allow the syntax suggested by you to be implemented.

File: commandapi-core/src/main/java/dev/jorel/commandapi/AbstractCommandAPICommand.java [hunk 2]
```diff
@@ -304,32 +304,14 @@ public Impl copy() {
...
-				if (i >= firstOptionalArgumentIndex) {
-					Argument[] arguments = (Argument[]) new AbstractArgument[i];
-					System.arraycopy(argumentsArray, 0, arguments, 0, i);
-					argumentsToRegister.add(arguments);
-				}
-			}
+		// Create a List of arrays that hold arguments to register split based on where optional arguments are
+		List<Argument> currentCommand = new ArrayList<>(argumentsArray.length);
+		for(Argument argument : argumentsArray) {
+			if(argument.isOptional()) argumentsToRegister.add((Argument[]) currentCommand.toArray(new AbstractArgument[0]));
```
[the remark this answers is not in the document]
Comment (DerEchtePilz) [reply commandapi-core/src/main/java/dev/jorel/commandapi/AbstractCommandAPICommand.java:310]: `Argument` is a type parameter so yes, we do have to use `new AbstractArgument[0]`

**Comment by DerEchtePilz** [comment]
Also, how would this system handle this:
```
List<ThingToTestCommandAPIThings> list = List.of(
    new ThingToTestCommandAPIThings(),
    new ThingToTestCommandAPIThings().setOptional(true),
    new ThingToTestCommandAPIThings().setOptional(true),
    new ThingToTestCommandAPIThings().setOptional(true),
    new ThingToTestCommandAPIThings().setOptional(false)
);
```
This was my way of testing that system without needing to import something and I am able to test this in a plain Java project.
Your new system generates this:
```
[ThingToTestCommandAPIThings{optional=false}]
[ThingToTestCommandAPIThings{optional=false}, ThingToTestCommandAPIThings{optional=true}]
[ThingToTestCommandAPIThings{optional=false}, ThingToTestCommandAPIThings{optional=true}, ThingToTestCommandAPIThings{optional=true}]
[ThingToTestCommandAPIThings{optional=false}, ThingToTestCommandAPIThings{optional=true}, ThingToTestCommandAPIThings{optional=true}, ThingToTestCommandAPIThings{optional=true}, ThingToTestCommandAPIThings{optional=false}]
```

However, I would expect it to only generate one result because I have a required argument at the end.
This was also part of the decision why I decided to even bother with those optional argument constraints.
In my opinion (and please don't take it personal, because it really isn't meant to offend anyone) we should stick with the system I implemented since there are now many more possibilities because optional arguments are implemented.

**Comment by willkroboth** [comment]
> Also, how would this system handle this:
> 
> ```
> List<ThingToTestCommandAPIThings> list = List.of(
>     new ThingToTestCommandAPIThings(),
>     new ThingToTestCommandAPIThings().setOptional(true),
>     new ThingToTestCommandAPIThings().setOptional(true),
>     new ThingToTestCommandAPIThings().setOptional(true),
>     new ThingToTestCommandAPIThings().setOptional(false)
> );
> ```
> 
> This was my way of testing that system without needing to import something and I am able to test this in a plain Java project. Your new system generates this:
> 
> ```
> [ThingToTestCommandAPIThings{optional=false}]
> [ThingToTestCommandAPIThings{optional=false}, ThingToTestCommandAPIThings{optional=true}]
> [ThingToTestCommandAPIThings{optional=false}, ThingToTestCommandAPIThings{optional=true}, ThingToTestCommandAPIThings{optional=true}]
> [ThingToTestCommandAPIThings{optional=false}, ThingToTestCommandAPIThings{optional=true}, ThingToTestCommandAPIThings{optional=true}, ThingToTestCommandAPIThings{optional=true}, ThingToTestCommandAPIThings{optional=false}]
> ```

Yeah, the output there is what I expected. Simplifying that a bit so its easier to read and talk about, the input is:
```
A, B (optional), C (optional), D (optional), E
```

And the output is:
```
A
A, B
A, B, C
A, B, C, D, E
```
So, I guess calling a required argument that is after an optional argument "required" isn't quite right, since you don't need to always include it (E is "required", but not present in all of the commands). That is necessary to avoid ambiguity while still being useful. With the example, if E always had to be included, you could either add it to the end of every command and get:
```
A, E
A, B, E
A, B, C, E
A, B, C, D, E
```
or require all arguments before it, basically ignoring their optional status. So yeah, you are correct that if E (the last argument) was "required" that wouldn't really work.

What my system is doing instead is actually pretty much linking the arguments. So E isn't strictly fully required, it is required if D is present. You could say E ***is*** linked to D, but without an explicitly linked argument structure, more implicitly by the fact that E is not optional (required) and after D.

I hope that explains why the output was correct to be like that? Maybe there is a better word than required for arguments like E and `rating` that would make this property clearer.

**Comment by DerEchtePilz** [comment]
> Yeah, the output there is what I expected. Simplifying that a bit so its easier to read and talk about, the input is:
> 
> ```
> A, B (optional), C (optional), D (optional), E
> ```
> 
> And the output is:
> 
> ```
> A
> A, B
> A, B, C
> A, B, C, D, E
> ```
> 
> So, I guess calling a required argument that is after an optional argument "required" isn't quite right, since you don't need to always include it (E is "required", but not present in all of the commands). That is necessary to avoid ambiguity while still being useful. With the example, if E always had to be included, you could either add it to the end of every command and get:
> 
> ```
> A, E
> A, B, E
> A, B, C, E
> A, B, C, D, E
> ```
> 
> or require all arguments before it, basically ignoring their optional status. So yeah, you are correct that if E (the last argument) was "required" that wouldn't really work.
> 
> What my system is doing instead is actually pretty much linking the arguments. So E isn't strictly fully required, it is required if D is present. You could say E _**is**_ linked to D, but without an explicitly linked argument structure, more implicitly by the fact that E is not optional (required) and after D.
> 
> I hope that explains why the output was correct to be like that? Maybe there is a better word than required for arguments like E and `rating` that would make this property clearer.

Well, what I think is this:
- First option: Keep my system. The reason for that is that we can pretty much do everything we want with a combination of subcommands and optional arguments
- Second option: Append `E` to every argument. We have the ability to access arguments by using their node names. So what we could do is throw the `OptionalArgumentException` if we have same node names on at least two arguments

File: commandapi-core/src/main/java/dev/jorel/commandapi/AbstractCommandAPICommand.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (willkroboth) [reply commandapi-core/src/main/java/dev/jorel/commandapi/AbstractCommandAPICommand.java:310]: Yea, Java doesn't like `new Argument[0]` -> Error: "Type parameter 'Argument' cannot be instantiated directly" :(

**Comment by willkroboth** [comment]
Note the example command:
```java
new CommandAPICommand("rate")
	.withArguments(
		new StringArgument("topic").setOptional(true),
		new IntegerArgument("rating", 0, 10),
		new PlayerArgument("target").setOptional(true)
	)
	.executes(info -> {
		String topic = (String) info.args().get("topic");
		if(topic == null) {
			info.sender().sendMessage("You didn't give a rating");
			return;
		}

		int rating = (int) info.args().get("rating");
		CommandSender target = (CommandSender) info.args().getOrDefault("target", info.sender());

		target.sendMessage("Your " + topic + " was rated: " + rating + "/10");
	})
	.register();
```

is equivalent to this CommandTree:
```java
new CommandTree("rate")
	.executes((info) -> info.sender().sendMessage("You didn't give a rating"))
	.then(
		new StringArgument("topic").then(
			new IntegerArgument("rating")
				.executes((info) -> {
					String topic = (String) info.args().get("topic");
					int rating = (int) info.args().get("rating");

					info.sender().sendMessage("Your " + topic + " was rated: " + rating + "/10");
				})
				.then(
					new PlayerArgument("target")
						.executes((info) -> {
							String topic = (String) info.args().get("topic");
							int rating = (int) info.args().get("rating");
							Player target = (Player) info.args().get("target");

							target.sendMessage("Your " + topic + " was rated: " + rating + "/10");
						})
				)
		)
	)
	.register();
```
I would consider the example command simpler and easier to understand, making this an example of when this sort of interaction between optional and required arguments would be useful.

File: commandapi-core/src/main/java/dev/jorel/commandapi/AbstractCommandAPICommand.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (JorelAli) [reply commandapi-core/src/main/java/dev/jorel/commandapi/AbstractCommandAPICommand.java:310]: Ah yes, rightly so. Didn't notice that. Thanks!

**Comment by JorelAli** [comment]
I think there's a bit of merit in the "linked" concept and I think it resolves this ambiguity in the declaration of the `/rate` command in the original post.

The claim is that the `/rate` command declared like this:

```java
new CommandAPICommand("rate")
    .withArguments(
        new StringArgument("topic").setOptional(true),
        new IntegerArgument("rating", 0, 10),
        new PlayerArgument("target").setOptional(true)
    )
```

leads to this list of possible commands:

```
/rate
/rate <topic> <rating>
/rate <topic> <rating> <target>
```

However, the ambiguity here is that this could also be interpreted as a valid command:

```
/rate <rating>
/rate <rating> <target>
```

The requirement that "rating" _must_ follow "topic" is 'lost' in this (the original) implementation. By 'lost', it may be inferred by the implementation of the CommandAPI, but as someone reading the code, it doesn't state this. The idea of having linked arguments addresses this - one (or more) argument(s) that must follow another argument and the whole thing can be classed as "an optional argument".

-----

I think the idea of some sort of "combined" argument makes sense. Say we have something like this:

```java
class CombinedArgument {
    CombinedArgument(Argument... args) { /* Some implementation */ }
}
```
```java
new CommandAPICommand("rate")
    .withArguments(
        new CombinedArgument(
            new StringArgument("topic").setOptional(true), // [1]
            new IntegerArgument("rating", 0, 10)
        ).setOptional(true),
        new PlayerArgument("target").setOptional(true)
    )
```

In this example, it's absolutely clear that "topic and rating go together". All properties of the inner arguments (e.g. `[1]` or setting permissions/requirements) are ignored, they must be implemented by the parent argument (the "combined argument"). We can then flag _the whole thing_ as "optional", but it's 100% clear what's going on.

The reason I think this works better than the original "linked" argument is because it avoids conflicting permissions on arguments - if you have a permission set on `rating` but _not_ on `topic`, then the user ends up with a half-baked command. Sure, you can implement this with linked arguments, but it's hard for a developer to figure out "which of the two arguments do I have to put my permission on to make it apply to both of them?"

-----

Briefly touching on some of the earlier conversation, in particular this example:

```
A, E
A, B, E
A, B, C, E
A, B, C, D, E
```

This was one of the primary reasons that the implementation of optional arguments stood around for 2 years. I initially wanted to have this sort of behaviour, but upon seeing the implementation by @DerEchtePilz, I found that it became a lot easier to interpret and design commands without having to worry about this. (Of course, being able to access arguments by their node name makes this significantly easier to deal with as well!)

**Comment by JorelAli** [comment]
I think we're going to sunset this PR in favour of #395. Linked/combined arguments seems to flow better and have less ambiguity.

**Pull request closed without merging**